lib/streamlit/runtime/scriptrunner/script_runner.py

```diff
@@ -522,43 +522,25 @@ def _run_script(self, rerun_data: RerunData) -> None:
-
-            # Saving and restoring our original cursors/dg_stack is needed
-            # specifically to handle the case where a RerunException is raised while
-            # running a fragment. In this case, we need to restore both to their states
-            # at the start of the script run to ensure that we write to the correct
-            # places in the app during the rerun (without this, ctx.cursors and dg_stack
-            # will still be set to the snapshots they were restored from when running
-            # the fragment).
-            original_cursors = ctx.cursors
-            original_dg_stack = dg_stack.get()
-
-            # If the script stops early, we don't want to remove unseen widgets,
-            # so we track this to potentially skip session state cleanup later.
-            premature_stop: bool = False
-
-            try:
-                # Create fake module. This gives us a name global namespace to
-                # execute the code in.
-                module = self._new_module("__main__")
-
-                # Install the fake module as the __main__ module. This allows
-                # the pickle module to work inside the user's code, since it now
-                # can know the module where the pickled objects stem from.
-                # IMPORTANT: This means we can't use "if __name__ == '__main__'" in
-                # our code, as it will point to the wrong module!!!
-                sys.modules["__main__"] = module
-
-                # Add special variables to the module's globals dict.
-                # Note: The following is a requirement for the CodeHasher to
-                # work correctly. The CodeHasher is scoped to
-                # files contained in the directory of __main__.__file__, which we
-                # assume is the main script directory.
-                module.__dict__["__file__"] = script_path
-
+            # Create fake module. This gives us a name global namespace to
+            # execute the code in.
+            module = self._new_module("__main__")
+
+            # Install the fake module as the __main__ module. This allows
+            # the pickle module to work inside the user's code, since it now
+            # can know the module where the pickled objects stem from.
+            # IMPORTANT: This means we can't use "if __name__ == '__main__'" in
+            # our code, as it will point to the wrong module!!!
+            sys.modules["__main__"] = module
+
+            # Add special variables to the module's globals dict.
+            # Note: The following is a requirement for the CodeHasher to
+            # work correctly. The CodeHasher is scoped to
+            # files contained in the directory of __main__.__file__, which we
+            # assume is the main script directory.
+            module.__dict__["__file__"] = script_path
+
+            def code_to_exec(code=code, module=module, ctx=ctx, rerun_data=rerun_data):
```

Comment:
nit: do we need to define parameters for this function with defaults bound to `code`/`module`/etc? I would guess that it should be fine to have this function take no parameters and allow all of the variables defined above be captured by the closure

Reply:
yeah I thought so too, but the linter complained and suggested this as a fix

![Screenshot 2024-06-13 at 12 13 36](https://github.com/streamlit/streamlit/assets/3775781/4b6deeb2-aad3-4f2a-aa93-2e474ca98d35)

See the [ruff docs](https://docs.astral.sh/ruff/rules/function-uses-loop-variable/) for why its bad.

I think our scenario is different because the code is called also directly within the loop, but perhaps it makes sense to stay with this pattern regardless to avoid issues when doing refactoring or something like that.  What do you think?